### ISSUE TITLE
Authentication for accessing Google PubSub (and other third-party APIs)

### ISSUE THREAD

**Issue by clearjs** [tony612/grpc-elixir#71]
Hi @tony612!

Thanks for making this library. 👍

I want to use it for accessing Google PubSub sevice. Right now I'm in the process of trying to figure out how to do the authentication properly. Your route_guide example uses a self-signed CA certificate in both client and server. Does this mean that I should find somewhere Google's CA cert and use that? After looking at the source code of gun adapter in grpc-elixir and gun docs, I tried specifying the following opts to create a `channel`:

```elixir
opts = [cred: GRPC.Credential.new(ssl: [])] # This is a workaround to make gun adapter establish a TLS connection without providing it a CA cert.
{:ok, channel} = GRPC.Stub.connect("pubsub.googleapis.com:443", opts) 
```

I'm not sure whether it works properly and how to verify that.

Next, I used goth to retrieve a JWT token. I verified that it works when accessing my Google PubSub project using https://github.com/GoogleCloudPlatform/elixir-google-api that works over REST.

Next, I generated Elixir code from Google PubSub proto files (for client stub and relevant messages). 

I also added ability to include the `authorization` header. I'll post a PR with that code soon.

However, I wasn't able to make this all work. When I execute the code below, I get 404.

```elixir
token = client.authenticate() # Retrieve a JWT token using goth.
project_id = "pubsub-elixir" # A project I created with Google PubSub.
req = Google.Pubsub.V1.ListTopicsRequest.new(project: "projects/#{project_id}", page_size: 5)
{_, reply} = channel |> Google.Pubsub.V1.Publisher.Stub.list_topics(req, token: token) # Right now this returns `%GRPC.RPCError{message: "status got is 404 instead of 200", status: 13}` as `reply`.
```

Please advise what should I do differently, or in which direction to move.

**Comment by clearjs** [comment]
I pushed a sample (not yet working) implementation here: https://github.com/clearjs/pubsub-grpc-auth-elixir. I tried an approach similar to https://medium.com/@KevinHoffman/mutual-tls-over-grpc-with-elixir-a071d514deb3.

Here are my new settings: https://github.com/clearjs/elixir-auth/blob/master/priv/client.exs.

I added CA certificates from https://github.com/certifi/erlang-certifi for channel authentication.

I also included a certificate and a private key files, although they don't seem to have any effect:

- generated GOOGLE_APPLICATION_CREDENTIALS JSON file with google console, saved value from its private_key field to a client-key.pem file, replacing '/n' with actual newline

- followed [the url](https://www.googleapis.com/robot/v1/metadata/x509/test-service-account%40pubsub-elixir.iam.gserviceaccount.com) from client_x509_cert_url field, and saved a certificate corresponding to the private_key_id field (60b7a01a580e8c38554e0c9ed1494791a6396798 in this case) to client-cert.pem file, also replacing '/n'

I still use goth for generating tokens for per-call authentication and put them to authorization header after "Bearer " prefix.

But when I execute the client code above, I'm getting the following error when server_name_indication is not set in ssl parameters:

[info] ['TLS', 32, 'client', 58, 32, 73, 110, 32, 115, 116, 97, 116, 101, 32, 'certify', 32, 'at ssl_handshake.erl:1335 generated CLIENT ALERT: Fatal - Handshake Failure - {bad_cert,hostname_check_failed}', 10]
{:error, "Error when opening connection: :timeout"}

If it is set to any value, only the last (timeout) error appears.

grpc swallows the following error:

<ins>That’s an error.</ins>\n  <p>The requested URL <code>/google.pubsub.v1.Publisher/ListTopics</code> was not found on this server.

which is a bit strange,  as I think that gRPC calls always go to the same URL, pubsub.googleapis.com, and only :path HTTP2 header should change. And the URL above looks correct. Maybe it is not being found because of incorrect authentication.

**Comment by tony612** [comment]
Based on the 404 error, it seems the TLS works? I think your guess("incorrect authentication") is right. I didn't use Google PubSub, but I guess you should use gRPC metadata(like custom headers in HTTP/1)  to pass the token. The way you tried to pass token is wrong
```
list_topics(req, token: token)
```
You can find examples in https://github.com/tony612/grpc-elixir/blob/master/interop/lib/interop/client.ex#L93

**Comment by ssboisen** [comment]
Did you ever manage to get your GRPC client for Google Cloud PubSub working @clearjs?

I had similar problems with Google Cloud Bigtable. With the emulator everything worked as it should but when targeting googleapis.com I got a 404. After debugging with Wireshark and finding the differences between an official bigtable client for nodejs and mine using grpc-elixir I found the problem.

It turns out the problem is the Content-Type set by default in grpc-elixir to `application/grpc+proto`, googleapis expect that to be `application/grpc`. You can change the content-type easily as grpc-elixir already supports it, just set `content_type: "application/grpc"` using the opts part of the service calls defined in the stub.

Thanks to @tony612 for creating grpc-elixir and protobuf-elixir!

**Comment by clearjs** [comment]
@ssboisen thanks for sharing that info. This task has been on hold since August till yesterday, so I'm going to continue soon. Do you have working code for Bigtable anywhere?

**Comment by clearjs** [comment]
@ssboisen thanks for sharing this! This task has been on hold until yesterday, but now I plan to finish it. I added `content_type` as you suggested, and it now works!

```elixir
project_id="..." # your project_id from Google PubSub
token="..." # token string from Goth
req = Google.Pubsub.V1.ListTopicsRequest.new(project: "projects/#{project_id}", page_size: 5)
Google.Pubsub.V1.Publisher.Stub.list_topics(req, token: token, content_type: "application/grpc")
```

I used my fork of grpc-elixir which takes the `token` param and uses it to set the authentication header. However, as I mentioned in https://github.com/tony612/grpc-elixir/pull/72#issuecomment-409125995, there's a better way to do authentication. I'll implement it and make a pull request soon.

@tony612 you're right that
> The way you tried to pass token is wrong
This way works only with my fork (clearjs/grpc-elixir), and it's better to implement it differently: pass a function that returns tokens when creating `%GRPC.Channel{}` and let it use that. There's no need for using custom headers, however. Just adding `authentication: Bearer <token>` header is enough.
